libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Security/SFUserAccountManager+URLHandlers.m

```diff
@@ -0,0 +1,188 @@
+/*
+ SFUserAccountManager+URLHandlers.m
+ SalesforceSDKCore
+ 
+ Created by Raj Rao on 9/25/17.
+ 
+ Copyright (c) 2017-present, salesforce.com, inc. All rights reserved.
+ 
+ Redistribution and use of this software in source and binary forms, with or without modification,
+ are permitted provided that the following conditions are met:
+ * Redistributions of source code must retain the above copyright notice, this list of conditions
+ and the following disclaimer.
+ * Redistributions in binary form must reproduce the above copyright notice, this list of
+ conditions and the following disclaimer in the documentation and/or other materials provided
+ with the distribution.
+ * Neither the name of salesforce.com, inc. nor the names of its contributors may be used to
+ endorse or promote products derived from this software without specific prior written
+ permission of salesforce.com, inc.
+ 
+ THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR
+ IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
+ FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+ CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
+ WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#import "SFUserAccountManager+Internal.h"
+#import "SFUserAccountManager+URLHandlers.h"
+#import "SFSDKOAuthClientConfig.h"
+#import "SFSDKOAuthClientContext.h"
+
+static NSString *const KSFStateParam = @"state";
+static NSString *const kSFErrorReasonParam = @"errorReason";
+static NSString *const kSFUserHintParam = @"user_hint";
+static NSString *const kSFAppNameParam = @"app_name";
+static NSString *const kSFAppDescParam = @"app_desc";
+static NSString *const kSFCallingAppUrlParam = @"calling_app_url";
+
+@implementation SFUserAccountManager (URLHandlers)
+
+- (BOOL)handleNativeAuthResponse:(NSURL *_Nonnull)appUrlResponse options:(NSDictionary *_Nullable)options {
+    //should return the shared instance for advanced auth
+    NSString *state = [appUrlResponse valueForParameterName:KSFStateParam];
+    NSString *key = [NSString stringWithFormat:@"%@-ADVANCED", state];
+    SFSDKOAuthClient *client = [self.oauthClientInstances objectForKey:key];
```

Comment:
Could we have a proper class for the oauth client instances (that would internally have the mutable dictionary)?
It would have a method to get a SFSDKOauthClient by state or by credentials and the building of the key from the state or the credentials would be done within the class.

Reply:
@wmathurin The SFSDKOAuthClient itself doesn't really care about the  # of instances used by a consuming class. In this case, the SFUserAccountManger would own that responsibility since it needs multiple instances of the client around (support multiple in-flight requests). Now if we wrap this functionality in another class are you suggesting having a SFSDKOAuthClientManager ? or  a SFSDKOAuthClientCache?

Reply:
I think we should encapsulate the management of those instances (and the key calculation) in a proper class - and retrieve them through proper method instead of having a generic dictionary and key calculation code spread out in various places. I'd prefer a dedicated class (e.g. SFSDKOAuthClientCache) and SFUserAccountManager would hold a reference to an instance of it.